test/metatx/ERC2771Forwarder.test.js

```diff
@@ -0,0 +1,431 @@
+const ethSigUtil = require('eth-sig-util');
+const Wallet = require('ethereumjs-wallet').default;
+const { getDomain, domainType } = require('../helpers/eip712');
+const { expectRevertCustomError } = require('../helpers/customError');
+
+const { constants, expectRevert, expectEvent, time } = require('@openzeppelin/test-helpers');
+const { expect } = require('chai');
+
+const ERC2771Forwarder = artifacts.require('ERC2771Forwarder');
+const CallReceiverMock = artifacts.require('CallReceiverMock');
+
+contract('ERC2771Forwarder', function (accounts) {
+  const [, refundReceiver, another] = accounts;
+
+  const tamperedValues = {
+    from: another,
+    to: another,
+    value: web3.utils.toWei('0.5'),
+    data: '0x1742',
+    deadline: 0xdeadbeef,
+  };
+
+  beforeEach(async function () {
+    this.forwarder = await ERC2771Forwarder.new('ERC2771Forwarder');
+
+    this.domain = await getDomain(this.forwarder);
+    this.types = {
+      EIP712Domain: domainType(this.domain),
+      ForwardRequest: [
+        { name: 'from', type: 'address' },
+        { name: 'to', type: 'address' },
+        { name: 'value', type: 'uint256' },
+        { name: 'gas', type: 'uint256' },
+        { name: 'nonce', type: 'uint256' },
+        { name: 'deadline', type: 'uint48' },
+        { name: 'data', type: 'bytes' },
+      ],
+    };
+
+    this.alice = Wallet.generate();
+    this.alice.address = web3.utils.toChecksumAddress(this.alice.getAddressString());
+
+    this.timestamp = await time.latest();
+    this.request = {
+      from: this.alice.address,
+      to: constants.ZERO_ADDRESS,
+      value: '0',
+      gas: '100000',
+      data: '0x',
+      deadline: this.timestamp.toNumber() + 60, // 1 minute
+    };
+    this.requestData = {
+      ...this.request,
+      nonce: (await this.forwarder.nonces(this.alice.address)).toString(),
+    };
+
+    this.forgeData = request => ({
+      types: this.types,
+      domain: this.domain,
+      primaryType: 'ForwardRequest',
+      message: { ...this.requestData, ...request },
+    });
+    this.sign = (privateKey, request) =>
+      ethSigUtil.signTypedMessage(privateKey, {
+        data: this.forgeData(request),
+      });
+
+    this.requestData.signature = this.sign(this.alice.getPrivateKey());
+  });
+
+  context('verify', function () {
+    context('with valid signature', function () {
+      beforeEach(async function () {
+        expect(await this.forwarder.nonces(this.requestData.from)).to.be.bignumber.equal(
+          web3.utils.toBN(this.requestData.nonce),
+        );
+      });
+
+      it('success', async function () {
+        expect(await this.forwarder.verify(this.requestData)).to.be.equal(true);
+      });
+
+      afterEach(async function () {
+        expect(await this.forwarder.nonces(this.requestData.from)).to.be.bignumber.equal(
+          web3.utils.toBN(this.requestData.nonce),
+        );
+      });
+    });
```

Comment:
Why not combine all this into a single test?

```suggestion
    it('with valid signature', async function () {
      expect(await this.forwarder.nonces(this.requestData.from)).to.be.bignumber.equal(
        web3.utils.toBN(this.requestData.nonce),
      );
      expect(await this.forwarder.verify(this.requestData)).to.be.equal(true);
      expect(await this.forwarder.nonces(this.requestData.from)).to.be.bignumber.equal(
        web3.utils.toBN(this.requestData.nonce),
      );
    });
```

Reply:
It's just the way it was before but no preference, I added your suggestion